flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,268 @@
+		List<InputGateDeploymentDescriptor> inputGates = new ArrayList<>(inputEdges.length);
+
+		for (ExecutionEdge[] edges : inputEdges) {
+			ShuffleDescriptor[] consumedPartitions = getConsumedPartitionShuffleDescriptors(edges, allowLazyDeployment);
+			// If the produced partition has multiple consumers registered, we
+			// need to request the one matching our sub task index.
+			// TODO Refactor after removing the consumers from the intermediate result partitions
+			int numConsumerEdges = edges[0].getSource().getConsumers().get(0).size();
+
+			int queueToRequest = subtaskIndex % numConsumerEdges;
+
+			IntermediateResult consumedIntermediateResult = edges[0].getSource().getIntermediateResult();
+			IntermediateDataSetID resultId = consumedIntermediateResult.getId();
+			ResultPartitionType partitionType = consumedIntermediateResult.getResultType();
+
+			inputGates.add(new InputGateDeploymentDescriptor(
+				resultId,
+				partitionType,
+				queueToRequest,
+				consumedPartitions,
+				location));
+		}
+
+		return inputGates;
+	}
+
+	private static ShuffleDescriptor[] getConsumedPartitionShuffleDescriptors(
+			ExecutionEdge[] edges,
+			boolean allowLazyDeployment) throws ExecutionGraphException {
+		ShuffleDescriptor[] shuffleDescriptors = new ShuffleDescriptor[edges.length];
+		// Each edge is connected to a different result partition
+		for (int i = 0; i < edges.length; i++) {
+			shuffleDescriptors[i] =
+				getConsumedPartitionShuffleDescriptor(edges[i], allowLazyDeployment);
+		}
+		return shuffleDescriptors;
+	}
+
+	private static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ExecutionEdge edge,
+			boolean allowLazyDeployment) throws ExecutionGraphException {
+		IntermediateResultPartition consumedPartition = edge.getSource();
+		Execution producer = consumedPartition.getProducer().getCurrentExecutionAttempt();
+
+		ExecutionState producerState = producer.getState();
+		Optional<ResultPartitionDeploymentDescriptor> consumedPartitionDescriptor =
+			producer.getProducedPartition(consumedPartition.getPartitionId());
+
+		ResultPartitionID consumedPartitionId = new ResultPartitionID(
+			consumedPartition.getPartitionId(),
+			producer.getAttemptId());
+
+		return getConsumedPartitionShuffleDescriptor(
+			consumedPartitionId,
+			consumedPartition.getResultType(),
+			consumedPartition.isConsumable(),
+			producerState,
+			allowLazyDeployment,
+			consumedPartitionDescriptor.orElse(null));
+	}
+
+	@VisibleForTesting
+	static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ResultPartitionID consumedPartitionId,
+			ResultPartitionType resultPartitionType,
+			boolean isConsumable,
+			ExecutionState producerState,
+			boolean allowLazyDeployment,
+			@Nullable ResultPartitionDeploymentDescriptor consumedPartitionDescriptor) throws ExecutionGraphException {
+		// The producing task needs to be RUNNING or already FINISHED
+		if ((resultPartitionType.isPipelined() || isConsumable) &&
```

Comment:
We have three cases here. I am not sure whether we could change the sequence of checking the conditions. My concern is that let the short easy case happen early, and focus on the complicated common case in the end. That means if the last two `else if` `else` clauses could be checked earlier in this case. I have not thought through it yet. Just an initial thought.